### ISSUE TITLE
Delay occurs while getting the size from the FormattedText

### ISSUE THREAD

**Issue by ghost** [dotnet/wpf#6642]
* .NET Core Version: .Net 6.0
* Windows version: Windows11- 22000.675
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes
* Is this bug related specifically to tooling in Visual Studio -No

 
 
 **Problem description:**
Delay Occurs while getting the size from the Formatted Text.
 
 **Expected behavior:**
 Delay should not be occured.
 **Minimal repro:**
1.Run the sample
2.Click the Initialize Formatted Text button.

Note : Delay occurs, button is in pressed state for a long time. 
Please refer the below code snippet and gif file.
```
private void Button_Click(object sender, RoutedEventArgs e)
{

    for (int i = 0; i <= 30000; i++)
    {
        FormattedText text = new FormattedText("Demo", System.Globalization.CultureInfo.CurrentCulture, FlowDirection.LeftToRight, new Typeface(FontFamily, new FontStyle(), FontWeights.Normal, FontStretch), 14, Brushes.Black);
        Size demo = new Size(text.Width, text.Height);
    }
    MessageBox.Show("Initialized");
}
```
![ScreenCapture_6-2-2022 4 50 39 PM](https://user-images.githubusercontent.com/92511047/171631346-830650aa-7450-4ae9-a067-270eed92d05e.gif)

**Comment by miloush** [comment]
@sethuramkumar I think this issue needs a bit of more information. Is the performance affecting a scenario in your application? Why would 30000 iterations be too few or too many, how many iterations would be reasonable to take the time it is taking now?

**Comment by sethuramkumar** [comment]
@miloush Yes this causes the performance issue in my application, I need to initialize a formatted text and use it in my custom control. 30000 iteration is minimum. As you can see in the gif, the button is in pressed state for a long time after clicking, The button's pressed state maintained until the iteration over. It seems like delay in the application.
Is this an expected behavior ?

**Comment by ygra** [comment]
> Is this an expected behavior ?

If you do a lot of work synchronously in the Click event handler, then yes, WPF has no way of updating anything and the button remains in the last drawn state. You can do the work in the background on another thread or schedule it via the dispatcher if you require the button to "pop up" again before your computation starts.

**Comment by sethuramkumar** [comment]
Hi @ygra , I have tried using the dispatcher and threads, still the button's state is not updated properly.

```
Application.Current.Dispatcher.BeginInvoke(DispatcherPriority.Normal, new Action(() =>
{
    for (int i = 0; i <= 30000; i++)
    {
        FormattedText text = new FormattedText("Demo", System.Globalization.CultureInfo.CurrentCulture, FlowDirection.LeftToRight, new Typeface(FontFamily, new FontStyle(), FontWeights.Normal, FontStretch), 14, Brushes.Black);
        Size demo = new Size(text.Width, text.Height);
    }
    MessageBox.Show("Initialized");

})); ;
```
Is there any way to resolve this delay?

**Comment by miloush** [comment]
@sethuramkumar check https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcherpriority?view=windowsdesktop-6.0 `Normal` has higher priority than `Render`

**Comment by sethuramkumar** [comment]
Hi @miloush 
Tried using the DispatcherPriority as Render, still the delay in button state is not resolved.

**Comment by miloush** [comment]
How are you assessing the delay? If I click the button using spacebar, with mouse away from it, and use `Normal`, the button stays blue while processing. With `Render` (or lower), it gets gray and then starts processing. If you use mouse, the button is blue on hover, the processing starts and then if you move mouse away, that enqueues another render, which will not execute until your processing is done (regardless of its priority).

Either way, dispatcher can either be rendering or doing other things.

If you want to do stuff simultaneously, you will have to either need to be pre-emptive and make sure you give dispatcher chance to process other operations every now and then, or, you need a completely separate thread like @ygra suggested. Good news for you is that you don't need a dispatcher for measuring `FormattedText`, so separate thread works fine. If it is just for a short task, you can use `Task`:

```C#
var typef = new Typeface(FontFamily, new FontStyle(), FontWeights.Normal, FontStretch);
Task.Run(() =>
{
    for (int i = 0; i <= 30000; i++)
    {
        FormattedText text = new FormattedText("Demo", System.Globalization.CultureInfo.CurrentCulture, FlowDirection.LeftToRight, typef, 14, Brushes.Black);
        Size demo = new Size(text.Width, text.Height);
    }
    MessageBox.Show("Initialized");
});
```

(Caveat is obviously you can't touch any objects that require Dispatcher from the other thread.)

**Comment by lindexi** [comment]
How about:

```csharp
   for (int i = 0; i <= 30000; i++)
    {
       await  Application.Current.Dispatcher.BeginInvoke(DispatcherPriority.Normal, new Action(() =>
       {

        FormattedText text = new FormattedText("Demo", System.Globalization.CultureInfo.CurrentCulture, FlowDirection.LeftToRight, new Typeface(FontFamily, new FontStyle(), FontWeights.Normal, FontStretch), 14, Brushes.Black);
        Size demo = new Size(text.Width, text.Height);
  


       })); 
  }
    MessageBox.Show("Initialized");
```

**Comment by vijayarasan** [comment]
Hi @miloush / @lindexi ,

I have checked your suggestion.  When using dispatcher the reported problem still occurs. But using the GetSize method with the help of GlyphRun. The performance improved. But we are using FormattedText to the renderer text in Custom control. So, the GlyphRun size is mismatched compared the FormattedText. This workaround does not resolve the exact case. Can you please suggest improve the performance using FormattedText?
